### ISSUE TITLE
Create Singularity strategy for HPC systems

### ISSUE THREAD

**Issue by jacg** [jacg/nain4#66]
Many more people are familiar with Docker than Nix.

Nix can generate docker images automatically: https://nixos.org/#asciinema-demo-example_4

IIRC, there is also (less well developed and maintained?) support for generating Singularity images.

**Comment by gonzaponte** [comment]
Will this help with the MacOS issues?

**Comment by jacg** [comment]
I don't know, because I don't know how docker images containing Linux-x86-64 binaries are executed on Apple Silicon.

When I searched around this topic a while ago, I got as far as seeing that there are issues with running docker images on Apple Silicon.

In principle, it *might*; in practice, *I* have no idea. I have avoided Docker like the plague since I started using Nix, which is before Apple Silicon was a thing.

**Comment by jacg** [comment]
Even if x86-64 Docker images could run reliably on Apple Silicon (and the internet tells me that they don't), they would suffer from some emulation overhead or other, degrading performance. Which we really want to avoid in something like Geant4.

So we'd want to build arm64 images. To what extent arm64 images built on Linux work reliably on Apple Silicon, I don't know, but given that that's the whole point of Docker, and that Docker on Apple Silicon is a thing, there should be no problem.

The next most obvious problem I can think of, is that our dependencies (prime suspect: Geant4) aren't supported in Nix on aarch64-linux. But if they are, we should be able to create both x86-64 and aarch64 Docker images, which *should* be a viable workaround for any MacOS problems, both on Intel and Apple Silicon.

**Comment by jacg** [comment]
Status update: it seems that nain4-via-Nix runs just fine on macOS, both Intel and Apple Silicon. So it seems that Docker is not needed to help things work on macOS.

Containerization is likely to be more useful on HPC systems where we don't have administrator rights. Because of the security implications of privilege escalation in Docker, you won't find many HPC systems allowing the use of Docker, but typically do allow Singularity:

+ Nix does have tools for generating Singularity containers, similar to those it has for generating Docker containers, but these are likely to be much less-well maintained than those for Docker. I have never tried them myself
+ It seems that Singularity itself has tools for using Docker images.

The above two points suggest it shouldn't be too difficult to get nain4 to run in a container on an HPC system that supports Singularity ... in theory.

However

+ When reading atlas' documentation I inferred that Singularity was available, but User 1 couldn't find any evidence of it being installed on the system.
+  Atlas documentation's own example on how to use SLURM, doesn't work correctly.

so *in practice* I doubt that it will be that simple.

In brief, we'll need to start taking a closer look at Singularity.

**Comment by jacg** [comment]
It looks like Singularity is now called Apptainer: https://apptainer.org/news/community-announcement-20211130

**Comment by jacg** [comment]
It looks like Singularity is now called Apptainer: https://apptainer.org/news/community-announcement-20211130.

And it comes pre-installed at CERN:

```
-bash-4.2$ hostname
lxplus724.cern.ch
-bash-4.2$ singularity --version
apptainer version 1.2.2-1.el7
```

**Comment by jacg** [comment]
Singularity containers are very easy to create from Nix, using `singularity-tools.buildImage`, although we had to explicitly ask for more `diskSize` and `memSize`  than the default when putting a `nain4` app inside one.

Such a container with `nixpkgs.hello` runs smoothly on the `singularity` (`apptainer`) that is pre-installed on CERN's lxplus, even though that is version 1.2.2 and we built the container with a Nix-provided version 3.11.1

However, there are problems

+ A nain4 app needs Geant4 along with various G4 data packages, and specific environment variables must be set in order for G4 to be able to find these data. `singularity-tools.buildImage` is completely undocumented, but looking at the source code, there doesn't seem to be anything that supports setting environment variables. The Singularity documentation does provide a way, using a Singularity definition file, but `singularity-tools` seems not to use this at all.

+ If you include all the G4 data packages, they weigh in at over 1GB. Nix automatically ensures that different derivations which depend on the same version of the data, share them. If we bundle these date in the containers, along with every n4 app that uses them, that's a huge waste of space.

Both of the above problems can be addressed by doing things by hand: copying G4 data to a shared location and setting environment variables by hand ... but then you lose all the comfort and guarantees that Nix gives you.

In this context, it seems even more worthwhile to give [`nix-portable`](https://github.com/DavHau/nix-portable) another go. We did get a user to try it on Atlas, and it failed. I don't recall the details.

`nix-portable` would address all of the above problems, though non-duplication of dependencies (such as the G4 data) would work on a per-user basis, whereas a proper installation of Nix would avoid duplication between all users on the whole system.

**Comment by jacg** [comment]
Proof of concept has been merged in db96a5bd124114c27f5f261f892207ec3c5969a2.

It has been tested by hand on

+ lxplus7.cern.ch (apptainer version 1.2.2-1.el7)
+ frontend1petalo (singularity version 3.8.0-1.el8)

Need someone with access to DIPC machines to test it there. I fear the worst ... I seem to recall that singularity wasn't installed there at all!

We have not managed to create an automated test on GHA, for lack of kvm support.
